aten/src/ATen/core/ivalue.h

```diff
@@ -457,19 +459,17 @@ struct CAFFE2_API IValue final {
 
 // Future
 struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
-  explicit Future(IValue result_) : result(result_), ready(true) {}
+  explicit Future(std::future<IValue>&& future_) : future(std::move(future_)) {}
 
-  IValue get() const {
-    AT_ASSERT(ready);
-    return result;
+  IValue get() {
```

Comment:
This patch is a good way of testing the multi-threaded perf before we add a cooperative thread pool. However, this call to `future.get()` is going to cause the caller to block if the future is not ready. We eventually want a cooperative threading model, where when this method is called by the interpreter, if the future is not ready, then the interpreter suspends itself, schedules a task to complete itself, and then returns a Future to caller of the interpreter that can be waited on to get the result of the interpreter. Calling `wait` on the future would pull the current thread into the group of worker threads until that Future is finished.

I think the next step is to implement this "interpreter suspend" behavior, even if we do not have a thread pool that can use it. That way the only remaining step is to integrate the threadpool.

Reply:
Thanks for the comments. Shall we make a stack of `(InterpreterState*, Stack*)` pair so that when the current interpreter suspends (i.e., `wait()`), it pushes its own `(InterpreterState*, Stack*)` to the stack and then the newly forked `(InterpreterState*, Stack*)`? A simple paradigm could be keeping calling `run()` for the top element of the stack till it is empty.
We can chat in person if that works for you.